nifi-extension-bundles/nifi-aws-bundle/nifi-aws-kinesis/src/test/java/org/apache/nifi/processors/aws/kinesis/ReaderRecordProcessorTest.java

```diff
@@ -0,0 +1,401 @@
+        assertFlowFileRecords(successFlowFile, records.get(0), records.get(2), records.get(4));
+
+        assertAll(result.parseFailureFlowFiles().stream().map(
+                failureFlowFile -> () -> {
+                    assertNotNull(failureFlowFile.getAttribute(RECORD_ERROR_MESSAGE));
+                    assertEquals(TEST_SHARD_ID, failureFlowFile.getAttribute(SHARD_ID));
+                }
+        ));
+    }
+
+    @Test
+    void testIOExceptionDuringReaderCreation() {
+        final RecordReaderFactory failingReaderFactory = new MockRecordParser() {
+            @Override
+            public RecordReader createRecordReader(Map<String, String> variables, InputStream in, long inputLength, ComponentLog logger) throws IOException {
+                throw new IOException("Failed to create reader");
+            }
+        };
+
+        final ReaderRecordProcessor processor = new ReaderRecordProcessor(failingReaderFactory, jsonWriter, logger);
+
+        final KinesisClientRecord record = createKinesisRecord("{\"name\":\"John\"}", "1");
+        final List<KinesisClientRecord> records = List.of(record);
+
+        final ProcessingResult result = processor.processRecords(session, TEST_STREAM_NAME, TEST_SHARD_ID, records);
+
+        assertEquals(0, result.successFlowFiles().size());
+        assertEquals(1, result.parseFailureFlowFiles().size());
+
+        final MockFlowFile failureFlowFile = (MockFlowFile) result.parseFailureFlowFiles().getFirst();
+        assertTrue(failureFlowFile.getAttribute(RECORD_ERROR_MESSAGE).contains("Failed to create reader"));
+        failureFlowFile.assertContentEquals(KinesisRecordPayload.extract(record), UTF_8);
+    }
+
+    @Test
+    void testMalformedRecordExceptionDuringReading() {
+        final ReaderRecordProcessor processor = new ReaderRecordProcessor(getMalformedRecordExceptionReader(), jsonWriter, logger);
+
+        final KinesisClientRecord record = createKinesisRecord("{\"name\":\"John\"}", "1");
+        final List<KinesisClientRecord> records = Collections.singletonList(record);
+
+        final ProcessingResult result = processor.processRecords(session, TEST_STREAM_NAME, TEST_SHARD_ID, records);
+
+        assertEquals(0, result.successFlowFiles().size());
+        assertEquals(1, result.parseFailureFlowFiles().size());
+
+        final MockFlowFile failureFlowFile = (MockFlowFile) result.parseFailureFlowFiles().getFirst();
+        assertTrue(failureFlowFile.getAttribute(RECORD_ERROR_MESSAGE).contains("Test exception"));
+        failureFlowFile.assertContentEquals(KinesisRecordPayload.extract(record), UTF_8);
+    }
+
+    @Test
+    void testInvalidRecordsWithSchemaEvolution() {
+        final ReaderRecordProcessor processor = new ReaderRecordProcessor(jsonReader, jsonWriter, logger);
+
+        final List<KinesisClientRecord> records = List.of(
+                createKinesisRecord("{\"name\":\"John\",\"age\":30}", "1"), // Schema A
```

Comment:
A number of these JSON strings are the same. It would be helpful to create static strings for the reused JSON record strings.